c2/c2-client-bundle/c2-client-base/src/main/java/org/apache/nifi/c2/client/C2ClientBase.java

```diff
@@ -37,39 +35,39 @@ public abstract class C2ClientBase implements C2Client {
     public C2ClientBase() {
         objectMapper = new ObjectMapper();
         objectMapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
+        objectMapper.setSerializationInclusion(JsonInclude.Include.NON_NULL);
     }
 
-    @Override
-    public C2HeartbeatResponse publishHeartbeat(C2Heartbeat heartbeat) throws IOException {
-        String heartbeatString = "";
-
-        if (heartbeat != null) {
-
-            final Payload payload = new Payload(heartbeat);
-            try {
-                heartbeatString = convertPayloadToString(payload);
-            } catch (IOException e) {
-                logger.info("Instance is currently restarting and cannot heartbeat.");
-                return null;
-            }
+    protected Optional<String> serialiseHeartbeat(C2Heartbeat heartbeat) {
```

Comment:
At this point this is more like a utility class supporting the serialization / deserialization

Reply:
Yea a possible alternative is to create another interface like `C2Serializer` or something with methods for serialize and deserialize using generics, like the deserialize method you added, and there could an implementation of `C2JacksonSerializer`. Then there would be no base client and the impls would use an instance of the serializer. I'm not sure if this is really a better design, unless you think there could ever be different serializations, it would just be favoring composition vs inheritance.

Reply:
I'm fine with this being a utility class, as I was moving code around I figured we'd have some shared functionality and put it in a client base, but if they really are just utilities like serialization, feel free to rename/move stuff around, if we don't need a client base then so be it.

Reply:
In the end I went with the generic C2Serializer approach (composition over inheritance). Also removed C2ClientBase which became empty after this change. As Matt wrote there could be use for it in the future. If it comes to that we can easily reintroduce it.